gensim/models/fasttext_inner.pyx

```diff
@@ -0,0 +1,544 @@
+        effective_sentences += 1
+        sentence_idx[effective_sentences] = effective_words
+
+        if effective_words == MAX_SENTENCE_LEN:
+            break  # TODO: log warning, tally overflow?
+
+    # precompute "reduced window" offsets in a single randint() call
+    for i, item in enumerate(model.random.randint(0, window, effective_words)):
+        reduced_windows[i] = item
+
+    # release GIL & train on all sentences
+    with nogil:
+        for sent_idx in range(effective_sentences):
+            idx_start = sentence_idx[sent_idx]
+            idx_end = sentence_idx[sent_idx + 1]
+            for i in range(idx_start, idx_end):
+                j = i - window + reduced_windows[i]
+                if j < idx_start:
+                    j = idx_start
+                k = i + window + 1 - reduced_windows[i]
+                if k > idx_end:
+                    k = idx_end
+
+                if hs:
+                    fast_sentence_cbow_hs(points[i], codes[i], codelens, neu1, syn0_vocab, syn0_ngrams, syn1, size, indexes,subwords_idx,
+                     subwords_idx_len,_alpha, work, i, j, k, cbow_mean, word_locks_vocab, word_locks_ngrams)
+                if negative:
+                    next_random = fast_sentence_cbow_neg(negative, cum_table, cum_table_len, codelens, neu1, syn0_vocab, syn0_ngrams, syn1neg, size, indexes, subwords_idx, 
+                        subwords_idx_len, _alpha, work, i, j, k, cbow_mean, next_random, word_locks_vocab, word_locks_ngrams)
+
+    return effective_words
+
+
+def init():
+    """
+    Precompute function `sigmoid(x) = 1 / (1 + exp(-x))`, for x values discretized
+    into table EXP_TABLE.  Also calculate log(sigmoid(x)) into LOG_TABLE.
+    """
+    global our_dot
+    global our_saxpy
+
+    cdef int i
+    cdef float *x = [<float>10.0]
+    cdef float *y = [<float>0.01]
+    cdef float expected = <float>0.1
+    cdef int size = 1
+    cdef double d_res
+    cdef float *p_res
+
+    # build the sigmoid table
+    for i in range(EXP_TABLE_SIZE):
+        EXP_TABLE[i] = <REAL_t>exp((i / <REAL_t>EXP_TABLE_SIZE * 2 - 1) * MAX_EXP)
+        EXP_TABLE[i] = <REAL_t>(EXP_TABLE[i] / (EXP_TABLE[i] + 1))
+        LOG_TABLE[i] = <REAL_t>log( EXP_TABLE[i] )
+
+    # check whether sdot returns double or float
+    d_res = dsdot(&size, x, &ONE, y, &ONE)
+    p_res = <float *>&d_res
+    if (abs(d_res - expected) < 0.0001):
+        our_dot = our_dot_double
+        our_saxpy = saxpy
+        return 0  # double
+    elif (abs(p_res[0] - expected) < 0.0001):
+        our_dot = our_dot_float
+        our_saxpy = saxpy
+        return 1  # float
+    else:
+        # neither => use cython loops, no BLAS
+        # actually, the BLAS is so messed up we'll probably have segfaulted above and never even reach here
+        our_dot = our_dot_noblas
+        our_saxpy = our_saxpy_noblas
+        return 2
+
+FAST_VERSION = init()  # initialize the module
+MAX_WORDS_IN_BATCH = MAX_SENTENCE_LEN
```

Comment:
No newline at the end of file.

Reply:
Some weird bug is causing the new line to be **not** reflected at the end of `.pyx` files even though my local commit shows it. Similar is the case with [`word2vec_inner.pyx`](https://github.com/RaRe-Technologies/gensim/blob/develop/gensim/models/word2vec_inner.pyx#L725).